sql/core/src/main/scala/org/apache/spark/sql/package.scala

```diff
@@ -98,7 +98,10 @@ package object sql {
       f
     } else {
       val st = Thread.currentThread().getStackTrace
-      var i = 3
+      var i = 0
```

Comment:
I'm not sure why we need this first loop to find the beginning of Spark code. We can be sure that the first stacktrace element belongs to `getStackTrace()`, the second belongs to `withOrigin()`, and because `withOrigin()` can't be called from user code the 3rd must belong to Spark code.

Reply:
Isn't it safe to don't assume any such things? BTW, this PR avoids any assumptions in `DataFrameQueryContext.apply`.

Reply:
Ok, I see your point. The link above is wrong as it points to a `Throwable.getStackTrace()`, but it seems the same weird behavior can happen with `Thread.getStackTrace()` as well: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Thread.html#getStackTrace()